test/cctest/test_environment.cc

```diff
@@ -0,0 +1,73 @@
+#include "gtest/gtest.h"
+#include "node.h"
+#include "env.h"
+#include "v8.h"
+#include "libplatform/libplatform.h"
+#include "node_test_fixture.h"
+
+using node::Environment;
+using node::IsolateData;
+using node::CreateIsolateData;
+using node::CreateEnvironment;
+using node::AtExit;
+using node::RunAtExit;
+
+static bool called_cb_1 = false;
+static bool called_cb_2 = false;
+static void at_exit_callback1(void* arg);
+static void at_exit_callback2(void* arg);
+static Environment* newenv(node::IsolateData* isolateData,
+                           v8::Local<v8::Context> context,
+                           char** argv);
+
+class EnvironmentTest : public NodeTestFixture {
+  virtual void TearDown() {
+    NodeTestFixture::TearDown();
+    called_cb_1 = false;
+    called_cb_2 = false;
+  }
+};
+
+TEST_F(EnvironmentTest, AtExitWithEnvironment) {
+  v8::HandleScope handle_scope(isolate_);
+  v8::Local<v8::Context> context = v8::Context::New(isolate_);
+  IsolateData* isolateData = CreateIsolateData(isolate_, uv_default_loop());
+  Argv argv{"node", "-e", ";"};
+  auto env = newenv(isolateData, context, *argv);
+
+  AtExit(env, at_exit_callback1);
+  RunAtExit(env);
+  EXPECT_TRUE(called_cb_1);
```

Comment:
Also call `FreeEnvironment()` and `FreeIsolateData()`.  Consider creating a RAII class for that or adding it to `EnvironmentTest`; the former is probably easier if you want to have more than one instance.